base/compiler/types.jl

```diff
@@ -212,3 +212,16 @@ may_compress(ni::NativeInterpreter) = true
 may_discard_trees(ni::NativeInterpreter) = true
 
 method_table(ai::AbstractInterpreter) = InternalMethodTable(get_world_counter(ai))
+
+# define inference bail out logic
+# `NativeInterpreter` bails out from inference when
+# - a lattice element grows up to `Any` (inter-procedural call, abstract apply)
+# - a lattice element gets down to `Bottom` (statement inference, local frame inference)
+# - inferring non-concrete toplevel call sites
+bail_out_call(interp::AbstractInterpreter, @nospecialize(t), sv)      = t === Any
+bail_out_apply(interp::AbstractInterpreter, @nospecialize(t), sv)     = t === Any
+bail_out_statement(interp::AbstractInterpreter, @nospecialize(t), sv) = t === Bottom
+bail_out_local(interp::AbstractInterpreter, @nospecialize(t), sv)     = t === Bottom
```

Comment:
What's the difference between `bail_out_statement` and `bail_out_local`?

Reply:
`bail_out_local` controls the frame-level bail out logic in `typeinf_local` while `bail_out_statement` does the statement-level bail out logic in `abstract_call_statement`.
The reason because I separate them is that it gives us more fine-grained control on the heuristic and we may want to use it.